server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAllocationHealthIndicatorService.java

```diff
@@ -0,0 +1,164 @@
+                status.addPrimary(shardRouting.primaryShard());
+                for (ShardRouting replicaShard : shardRouting.replicaShards()) {
+                    status.addReplica(replicaShard);
+                }
+            }
+        }
+
+        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails());
+    }
+
+    private static class ShardAllocationCounts {
+        private int unassigned = 0;
+        private int unassigned_restarting = 0;
+        private int initializing = 0;
+        private int started = 0;
+        private int relocating = 0;
+
+        public void increment(ShardRouting routing) {
+            switch (routing.state()) {
+                case UNASSIGNED -> {
+                    unassigned++;
+                    if (routing.unassignedInfo() != null && routing.unassignedInfo().getReason() == UnassignedInfo.Reason.NODE_RESTARTING) {
+                        unassigned_restarting++;
+                    }
+                }
+                case INITIALIZING -> initializing++;
+                case STARTED -> started++;
+                case RELOCATING -> relocating++;
+            }
+        }
+    }
+
+    private static class ShardAllocationStatus {
+        private final ShardAllocationCounts primaries = new ShardAllocationCounts();
+        private final ShardAllocationCounts replicas = new ShardAllocationCounts();
+
+        public void addPrimary(ShardRouting routing) {
+            primaries.increment(routing);
+        }
+
+        public void addReplica(ShardRouting routing) {
+            replicas.increment(routing);
+        }
+
+        public HealthStatus getStatus() {
+            if (primaries.unassigned > 0) {
+                return RED;
+            } else if (replicas.unassigned > replicas.unassigned_restarting) {
+                return YELLOW;
+            } else {
+                return GREEN;
+            }
+        }
+
+        public String getSummary() {
+            var builder = new StringBuilder("This cluster has ");
+            if (primaries.unassigned > 0 || replicas.unassigned > 0) {
+
+                builder.append(
+                    Stream.of(
+                        Stream.of("1 unassigned primary").filter(it -> primaries.unassigned == 1),
+                        Stream.of(primaries.unassigned + " unassigned primaries").filter(it -> primaries.unassigned > 1),
+                        Stream.of("1 unassigned replica").filter(it -> replicas.unassigned == 1),
+                        Stream.of(replicas.unassigned + " unassigned replicas").filter(it -> replicas.unassigned > 1)
+                    ).flatMap(Function.identity()).collect(joining(" and "))
+                ).append(".");
+            } else {
+                builder.append("no unassigned shards.");
+            }
+            return builder.toString();
+        }
+
+        public SimpleHealthIndicatorDetails getDetails() {
```

Comment:
If we are going to including the restarting logic, I think we need to include those in the details. "unavailable_restarting_replicas" or "restarting_replicas" seem fine.